### PR TITLE
Implement basic process number limiting closes #38

### LINKED ISSUE #38: Runtime: command should be able to spawn a limited # of processes
The runtime should limit how many processes that can be created when the `RunCmd` function is used. Currently, the only limitation put on the code is how long it runs.

The Linux syscall `setrlimit` will be useful, exposed in `os.syscall` package with [Setrlimit](https://pkg.go.dev/syscall#Setrlimit). Linux utility [prlimit](https://man7.org/linux/man-pages/man1/prlimit.1.html) (used to get and set resource limits for a process) will likely be most helpful in debugging since it is a wrapper for `setrlimit`. **NOTE** this will require being run in a Linux box or at least a Linux Docker container.

Example:

```
runner@runner-vm:~$ prlimit --pid $$ --nproc=10:12
runner@runner-vm:~$ ps
    PID TTY          TIME CMD
  50411 pts/0    00:00:00 bash
  50438 pts/0    00:00:00 ps
...
runner@runner-vm:~$ sleep 100 &
[6] 50444
runner@runner-vm:~$ sleep 100 &
-bash: fork: retry: Resource temporarily unavailable
```

#### More Resources

RLIMITs we care about described here: [getrlimit manpage](https://man7.org/linux/man-pages/man2/getrlimit.2.html)
Process Resource Limit: [prlimit syscall](https://linux.die.net/man/2/prlimit)
Go wrapper for syscall: [Prlimit](https://pkg.go.dev/golang.org/x/sys/unix#Prlimit)
Golang package/wrapper for these syscalls: [golang.org/x/sys/unix](https://pkg.go.dev/golang.org/x/sys/unix#section-readme)

Constants where `RLIMIT`s are defined: [pkg-constants](https://pkg.go.dev/golang.org/x/sys/unix?utm_source=godoc#pkg-constants)

AC:

- [x] runtime can set hard limits on how many processes can be spawned from a `RunCmd`
- [x] unit tests verify processes are killed when over nproc limit
- [x] fork new child process before applying limits

### PR DESCRIPTION
**Pull request by camerondurham**
#### Description

This PR implements #38 by adding a new binary `process` to handle setting limits in a separate process from the server process by using `setrlimit` syscalls (ref: [`getrlimit`](https://man7.org/linux/man-pages/man3/getrlimit.3p.html) for description). Have tested locally and with the `engine/main.go` test file but manually adding the build directory to the PATH is needed. See the Makefile `test` target for how to do this.

#### Details

The `prlimit`/`setrlimit` mechanism being used will apply limits on the process *if* the user is not root and these limits will be for that user.

> ... prlimit --nproc a.k.a. RLIMIT_NPROC a.k.a. ulimit -u is the maximum number of processes¹ for the user as a whole. If you already have 20 processes running, and you set the limit to 20, you can't create any new process. What matters is how many processes are running as your user, it doesn't matter who their parent is or what their limit setting is.
[[source](https://unix.stackexchange.com/a/303252/446190)]

See notes in this issue for more details: https://github.com/camerondurham/runner/issues/42#issuecomment-1054014347

Additional work is required to create or reuse a set of users on the server when making each request: we do not want multiple requests mapped to the same user since this means we cannot guarantee the users will have equal time/access to these resources (i.e. one can easily hog all the processes and other user's program will fail).

An "external" process may not be needed and hopefully can find a way around this. Will need to do additional testing if this external process is really needed or if we simply need a new non-root user. Plan to use `UidMappings` in [syscall.SysProcAttr](https://pkg.go.dev/syscall#SysProcAttr) to set user ids to uid != 0. Another example [here](https://stackoverflow.com/a/21706757/4676641).

Next steps:

- [x] create issues to run each runner process as a non-root user
- [x] ~~test if limiting behavior can be accomplished in same process~~ (tracking in #48)
- [x] create new issue to try limiting using cgroups instead of rlimits

Creating this PR to track progress. Will update and create new issues as needed.

### CODE REVIEW

**Comment by camerondurham** [comment]
### Manual Verification

Script run for testing purposes will need 3 total processes, since `&` on Unix-like systems will run in background as another process and then execute next command.
```bash
#!/bin/bash
echo hello world
sleep 1 &
sleep 1 &
echo i am done
```

```bash
# this runs shell script with limit of 3 processes (the script spawns 3 processes, so this limit is sufficient)
PROC_DEBUG=1 go run main.go -uid=1234 -gid=1234 -nprocs=3 -fsize=20000 -timeout=5 -cmd=$PWD/test.sh
parsed args: nprocs="3" fsize="20000" timeout:"5"  uid:"1234" gid:"1234" runcmd:"/home/engine/process/test.sh" cmds:[]
hello world
i am done


# this runs the script without enough processes, so we see "resource unavailable"
PROC_DEBUG=1 go run main.go -uid=1234 -gid=1234 -nprocs=2 -fsize=20000 -timeout=5 -cmd=$PWD/test.sh
hello world
/home/engine/process/test.sh: fork: retry: Resource temporarily unavailable
i am done
```

Recording: 

[asciicast](https://asciinema.org/a/bIG5Rdcuv9owjsMLwsIitStE3)

**Comment by camerondurham** [comment]
Merging since this was manually verified and documented. Will update README and any related breakage in new PR.

File: engine/runtime/runtime_test.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ func Test_RunCmd(t *testing.T) {
 		props *RunProps
 	}
 
-	runtimeAgent := NewTimeoutRuntime("test")
+	runtimeAgent := NewTimeoutRuntime("test", &NilProvider{})
```
Comment (siwei-li) [inline engine/runtime/runtime_test.go:14]: So we're not testing the limits set on runtime yet, correct?

Comment (camerondurham) [reply engine/runtime/runtime_test.go:14]: Correct, no testing of the limits at least in these unit test files. I've only manually tested limiting behavior and will create an issue to do automated end-to-end tests.

**Review by siwei-li — COMMENTED** [review]
I went through all the code and there's just one question: are we using engine/process/main.go as a standalone CLI and this is just how we test the limiter temporarily? Thank you

**Comment by camerondurham** [comment]
@siwei-li 
> I went through all the code and there's just one question: are we using engine/process/main.go as a standalone CLI and this is just how we test the limiter temporarily? Thank you

Yes, for now engine/process/main.go will be a standalone CLI on the server used to run user code in a new process. I think a new process (calling the CLI from the server's request handling thread) probably isn't necessary but need to do tests to verify.


But this temporary workflow is definitely not ideal. It's overly complicated and inefficient, just passing these rlimit values from the coderunner to the runtime to the CLI. Definitely need to improve that interface.

**Pull request merged**